syntax = "proto3";

service Emulator {
    rpc RecordEmulatorActions (EmulatorActions) returns (NextAction);
}

message NextAction {
    optional int32 num_instructions = 1;
}

message EmulatorActions {
    repeated Instruction instructions = 1;

    message MemoryAccess {
        uint64 vaddr = 1;
        uint32 size = 2;
        bool is_store = 3;
    }

    message BranchFlow {
        bool is_mispredict = 1;
    }

    message Instruction {
        bytes opcode = 1;
        uint64 addr = 2;
        uint32 size = 3;
        optional MemoryAccess memory_access = 4;
        optional BranchFlow branch_flow = 5;
    }
}
